File: web/src/types/classification.ts
export type TumorClass = 'glioma' | 'meningioma' | 'pituitary' | 'healthy'

export const TUMOR_CLASSES: TumorClass[] = ['glioma', 'meningioma', 'pituitary', 'healthy']

/**
 * Metadatos del modelo que está sirviendo, declarados por el propio artefacto.
 *
 * El preproceso NO se escribe a mano en la interfaz: lo declara el modelo
 * empaquetado y la API lo expone en `GET /api/v1/meta`. Ver patrón «Transform».
 */
export interface ActiveModelInfo {
  modelVersion: string
  preprocessLabel: string
  simulatedInference: boolean
}

/**
 * Métodos de explicabilidad que el artefacto puede declarar.
 *
 * Hoy solo hay uno: sensibilidad por oclusión. El modelo se sirve como ONNX y
 * `onnxruntime` no calcula gradientes, así que Grad-CAM no es implementable
 * sobre el artefacto real; la oclusión responde a la misma pregunta usando
 * únicamente pasadas hacia adelante.
 */
export type ExplanationMethod = 'occlusion'

/**
 * Justificación espacial de una predicción, producida por el propio modelo.
 *
 * Igual que el preproceso, el método NO se reimplementa en el cliente: la UI
 * solo transporta y pinta lo que el artefacto devuelve. Ver patrón
 * «Explainable Predictions».
 */
export interface PredictionExplanation {
  /** Identificador estable del método, tal y como lo declara el artefacto. */
  method: ExplanationMethod
  /** Etiqueta legible con los parámetros reales, p. ej. «Oclusión 32 px · paso 16 px». */
  methodLabel: string
  /**
   * Mapa de influencia normalizado a [0,1] como imagen (data URI).
   *
   * La intensidad viaja en el canal alfa, de modo que la imagen sirve
   * directamente de máscara CSS sobre la MRI, sin decodificar ni recolorear.
   */
  influenceMapDataUri: string
}

export interface ClassificationResult {
  predictedClass: TumorClass
  confidenceByClass: Record<TumorClass, number>
  description: string
  modelVersion: string
  preprocess: string
  countryCode: string
  /** Ausente cuando la petición no pidió explicación: cuesta N inferencias extra. */
  explanation: PredictionExplanation | null
}

File: api/app/schemas/classification.py
"""Esquemas de clasificación e información del modelo activo."""

from __future__ import annotations

from typing import Literal

from app.schemas.common import EsquemaBase

ClaseTumor = Literal["glioma", "meningioma", "pituitary", "healthy"]


class InfoModeloActivo(EsquemaBase):
    """Lo que la interfaz muestra sobre el modelo que está sirviendo.

    El preproceso lo declara el artefacto (patrón «Transform»); la API sólo lo
    transporta, no lo escribe a mano.
    """

    model_version: str
    preprocess_label: str
    simulated_inference: bool


class ExplicacionPrediccion(EsquemaBase):
    method: Literal["occlusion"]
    method_label: str
    influence_map_data_uri: str


class ResultadoClasificacion(EsquemaBase):
    predicted_class: ClaseTumor
    confidence_by_class: dict[str, float]
    description: str
    model_version: str
    preprocess: str
    country_code: str
    explanation: ExplicacionPrediccion | None = None


class InfoApi(EsquemaBase):
    """Respuesta de `GET /api/v1/meta`, que alimenta el indicador de estado."""

    app_name: str
    api_version: str
    environment: str
    data_source: str
    simulated_inference: bool
    model_version: str
    preprocess_label: str
